PR: https://github.com/odoo/odoo/pull/198141

From: 088cef5c2a6fc2e3ec139471f08418387636cc8c
From: Yagnik ✗ (yagp)
Date: 2025-08-11 05:23:37

Breaking data model changes scores: del:7 + add:0, change matches:
-    kanban_state_label = fields.Char(
-    legend_blocked = fields.Char(related='stage_id.legend_blocked', string='Kanban Blocked Explanation', readonly=True)
-    legend_done = fields.Char(related='stage_id.legend_done', string='Kanban Valid Explanation', readonly=True)
-    legend_normal = fields.Char(related='stage_id.legend_normal', string='Kanban Ongoing Explanation', readonly=True)
-    legend_blocked = fields.Char(
-    legend_done = fields.Char(
-    legend_normal = fields.Char(

Total Changes: 59

[IMP] {website_}event: make event cancelation easier

Purpose:
========
* Make it easier to cancel events.
* Make sure we do not send reminders for events that are over/canceled
as otherwise attendees may show up.

Specs:
======
* Add new event kanban_state `Cancel`
* Change the color of 'Blocked' state to orange and disable the
state configuration (Same as Projects/Task)
* On Cancelled Event :
  > Event should still appear on `/event`  with a 'Cancelled' banner in grid view
and cancelled label in list view
  > Registration on these event should be disabled
* Scheduled communications should not be triggered if the event state is cancelled
* Improve list view in `/event` page

Task-4197048

closes odoo/odoo#198141

Related: odoo/upgrade#7249
Signed-off-by: Thibault Delavallee (tde) <tde@openerp.com>

================================= pseudo patch: =================================

--- a/addons/event/models/event_event.py
+++ b/addons/event/models/event_event.py
@@ -94,16 +94,15 @@ class EventEvent(models.Model):
     # properties
     registration_properties_definition = fields.PropertiesDefinition('Registration Properties')
     # Kanban fields
-    kanban_state = fields.Selection([('normal', 'In Progress'), ('done', 'Done'), ('blocked', 'Blocked')], default='normal', copy=False)
-    kanban_state_label = fields.Char(
-        string='Kanban State Label', compute='_compute_kanban_state_label',
-        store=True, tracking=True)
+    kanban_state = fields.Selection([
+        ('normal', 'In Progress'),
+        ('done', 'Ready for Next Stage'),
+        ('blocked', 'Blocked'),
+        ('cancel', 'Cancelled')
+    ], default='normal', copy=False, compute='_compute_kanban_state', readonly=False, store=True, tracking=True)
     stage_id = fields.Many2one(
         'event.stage', ondelete='restrict', default=_get_default_stage_id,
         group_expand='_read_group_expand_full', tracking=True, copy=False)
-    legend_blocked = fields.Char(related='stage_id.legend_blocked', string='Kanban Blocked Explanation', readonly=True)
-    legend_done = fields.Char(related='stage_id.legend_done', string='Kanban Valid Explanation', readonly=True)
-    legend_normal = fields.Char(related='stage_id.legend_normal', string='Kanban Ongoing Explanation', readonly=True)
     # Seats and computation
     seats_max = fields.Integer(
         string='Maximum Attendees',
@@ -142,7 +141,7 @@ class EventEvent(models.Model):
     event_registrations_open = fields.Boolean(
         'Registration open', compute='_compute_event_registrations_open', compute_sudo=True,
         help="Registrations are open if:\n"
-        "- the event is not ended\n"
+        "- the event is not ended or not cancelled\n"
         "- there are seats available on event\n"
         "- the tickets are sellable (if ticketing is used)")
     event_registrations_sold_out = fields.Boolean(
@@ -249,16 +248,6 @@ class EventEvent(models.Model):
                 'event_type_id': False,
             })
 
-    @api.depends('stage_id', 'kanban_state')
-    def _compute_kanban_state_label(self):
-        for event in self:
-            if event.kanban_state == 'normal':
-                event.kanban_state_label = event.stage_id.legend_normal
-            elif event.kanban_state == 'blocked':
-                event.kanban_state_label = event.stage_id.legend_blocked
-            else:
-                event.kanban_state_label = event.stage_id.legend_done
-
     @api.depends('event_slot_count', 'is_multi_slots', 'seats_max', 'registration_ids.state', 'registration_ids.active')
     def _compute_seats(self):
         """ Determine available, reserved, used and taken seats. """
@@ -309,6 +298,7 @@ class EventEvent(models.Model):
     def _compute_event_registrations_open(self):
         """ Compute whether people may take registrations for this event
 
+          * for cancelled events, registrations are not open;
           * event.date_end -> if event is done, registrations are not open anymore;
           * event.start_sale_datetime -> lowest start date of tickets (if any; start_sale_datetime
             is False if no ticket are defined, see _compute_start_sale_date);
@@ -319,7 +309,8 @@ class EventEvent(models.Model):
             event = event._set_tz_context()
             current_datetime = fields.Datetime.context_timestamp(event, fields.Datetime.now())
             date_end_tz = event.date_end.astimezone(pytz.timezone(event.date_tz or 'UTC')) if event.date_end else False
-            event.event_registrations_open = event.event_registrations_started and \
+            event.event_registrations_open = event.kanban_state != 'cancel' and \
+                event.event_registrations_started and \
                 (date_end_tz >= current_datetime if date_end_tz else True) and \
                 (not event.seats_limited or not event.seats_max or event.seats_available) and \
                 (
@@ -560,6 +551,12 @@ class EventEvent(models.Model):
             if event.event_type_id and not is_html_empty(event.event_type_id.note):
                 event.note = event.event_type_id.note
 
+    @api.depends('stage_id')
+    def _compute_kanban_state(self):
+        for task in self:
+            if task.kanban_state != 'cancel':
+                task.kanban_state = 'normal'
+
     @api.depends('event_type_id')
     def _compute_ticket_instructions(self):
         for event in self:
@@ -642,12 +639,6 @@ class EventEvent(models.Model):
                     }
                 }
 
-    def write(self, vals):
-        if 'stage_id' in vals and 'kanban_state' not in vals:
-            # reset kanban state when changing stage
-            vals['kanban_state'] = 'normal'
-        return super().write(vals)
-
     @api.depends('event_registrations_sold_out', 'seats_limited', 'seats_max', 'seats_available')
     @api.depends_context('name_with_seats_availability')
     def _compute_display_name(self):

--- a/addons/event/models/event_mail.py
+++ b/addons/event/models/event_mail.py
@@ -62,7 +62,7 @@ class EventMail(models.Model):
         help='Slot-based communication')
     mail_done = fields.Boolean("Sent", copy=False, readonly=True)
     mail_state = fields.Selection(
-        [('running', 'Running'), ('scheduled', 'Scheduled'), ('sent', 'Sent'), ('error', 'Error')],
+        [('running', 'Running'), ('scheduled', 'Scheduled'), ('sent', 'Sent'), ('error', 'Error'), ('cancelled', 'Cancelled')],
         string='Global communication Status', compute='_compute_mail_state')
     mail_count_done = fields.Integer('# Sent', copy=False, readonly=True)
     notification_type = fields.Selection([('mail', 'Mail')], string='Send', compute='_compute_notification_type')
@@ -84,12 +84,15 @@ class EventMail(models.Model):
         if next_schedule and (cron := self.env.ref('event.event_mail_scheduler', raise_if_not_found=False)):
             cron._trigger(next_schedule)
 
-    @api.depends('error_datetime', 'interval_type', 'mail_done')
+    @api.depends('error_datetime', 'interval_type', 'mail_done', 'event_id')
     def _compute_mail_state(self):
         for scheduler in self:
             # issue detected
             if scheduler.error_datetime:
                 scheduler.mail_state = 'error'
+            # event cancelled
+            elif not scheduler.mail_done and scheduler.event_id.kanban_state == 'cancel':
+                scheduler.mail_state = 'cancelled'
             # registrations based
             elif scheduler.interval_type == 'after_sub':
                 scheduler.mail_state = 'running'
@@ -471,6 +474,8 @@ class EventMail(models.Model):
         schedulers = self.search([
             # skip archived events
             ('event_id.active', '=', True),
+            # skip if event is cancelled
+            ('event_id.kanban_state', '!=', 'cancel'),
             # scheduled
             ('scheduled_date', '<=', fields.Datetime.now()),
             # event-based: todo / attendee-based: running until event is not done

--- a/addons/event/models/event_stage.py
+++ b/addons/event/models/event_stage.py
@@ -16,12 +16,3 @@ class EventStage(models.Model):
     pipe_end = fields.Boolean(
         string='End Stage', default=False,
         help='Events will automatically be moved into this stage when they are finished. The event moved into this stage will automatically be set as green.')
-    legend_blocked = fields.Char(
-        'Red Kanban Label', default=lambda s: s.env._('Blocked'), translate=True, prefetch='legend', required=True,
-        help='Override the default value displayed for the blocked state for kanban selection.')
-    legend_done = fields.Char(
-        'Green Kanban Label', default=lambda s: s.env._('Ready for Next Stage'), translate=True, prefetch='legend', required=True,
-        help='Override the default value displayed for the done state for kanban selection.')
-    legend_normal = fields.Char(
-        'Grey Kanban Label', default=lambda s: s.env._('In Progress'), translate=True, prefetch='legend', required=True,
-        help='Override the default value displayed for the normal state for kanban selection.')
